File: ios/Classes/FlutterTencentUploadPlugin.m
#import "FlutterTencentUploadPlugin.h"
#import "TXUGCPublish.h"


@implementation FlutterTencentUploadPlugin {
    FlutterEventSink _eventSink;
    TXUGCPublish *_txUgcPublish;
    FlutterResult _result;
}

+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
    FlutterMethodChannel* channel = [FlutterMethodChannel
                                     methodChannelWithName:@"flutter_tencent_upload" binaryMessenger:[registrar messenger]];
    FlutterTencentUploadPlugin* instance = [[FlutterTencentUploadPlugin alloc] init];
    [registrar addMethodCallDelegate:instance channel:channel];
    
    FlutterEventChannel *event = [FlutterEventChannel eventChannelWithName:@"flutter_tencent_upload_stream" binaryMessenger:[registrar messenger]];
    [event setStreamHandler:instance];
}

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
    if ([@"init" isEqualToString:call.method]) {
        result([@"iOS " stringByAppendingString:[[UIDevice currentDevice] systemVersion]]);
    } else if ([@"uploadVideo" isEqualToString:call.method]) {
        if (call.arguments != nil && [call.arguments isKindOfClass:NSDictionary.class]) {
            _result = result;
            @try {
                if (_eventSink != nil){
                    _eventSink(@(0));
                }
                NSString *token = call.arguments[@"token"];
                NSString *videoPath = call.arguments[@"videoPath"];
                NSString *coverPath = call.arguments[@"coverPath"];
                TXUGCPublish *publish = [[TXUGCPublish alloc] init];
                publish.delegate = self;
                TXPublishParam *param = [[TXPublishParam alloc] init];
                param.signature = token;
                param.videoPath = videoPath;//[videoPath stringByReplacingOccurrencesOfString:@"file://" withString:@""];
                param.coverPath = coverPath;
                param.enableHTTPS = YES;
                [publish publishVideo:param];
            } @catch (NSException *exception) {
                NSLog(@"FlutterTencentUploadPlugin exception1: %@",exception);
                result(nil);
            }
        }
    } else {
        result(FlutterMethodNotImplemented);
    }
}
 

- (FlutterError * _Nullable)onCancelWithArguments:(id _Nullable)arguments {
    _eventSink = nil;
    return nil;
}

- (FlutterError * _Nullable)onListenWithArguments:(id _Nullable)arguments eventSink:(nonnull FlutterEventSink)events {
    _eventSink = events;
    return nil;
}
    

/**
 * 短视频发布进度
 */
- (void)onPublishProgress:(NSInteger)uploadBytes totalBytes: (NSInteger)totalBytes {
    if (_eventSink != nil){
        double progress = uploadBytes*1.0/totalBytes;
//        NSLog(@"uploadBytes:%zd, totalBytes:%zd, progress:%f", uploadBytes, totalBytes, progress);
        _eventSink(@(progress));
    }
}

/**
 * 短视频发布完成
 */
- (void)onPublishComplete:(TXPublishResult*)result {
    @try {
        if (result.retCode == 0) {
            NSDictionary *data = @{
                @"code": @1,
                @"msg": @"上传成功",
                @"data": @{
                    @"id": result.videoId,
                    @"url": result.videoURL,
                    @"coverUrl": result.coverURL,
                }
            };
            _result(data);
        } else {
            NSDictionary *data = @{
                @"code": @(result.retCode),
                @"msg": result.descMsg,
                @"data": @{},
            };
            _result(data);
        }
    } @catch(NSException *exception) {
        NSLog(@"FlutterTencentUploadPlugin exception1: %@",exception);
        _result(nil);
    }
    
}

/**
 * 短视频发布事件通知
 */
- (void)onPublishEvent:(NSDictionary*)evt {
    ;
}


@end

File: android/src/main/java/com/example/flutter_tencent_upload/FlutterTencentUploadPlugin.java
package com.example.flutter_tencent_upload;

import android.content.Context;

import androidx.annotation.NonNull;

import com.example.flutter_tencent_upload.videoupload.TXUGCPublish;
import com.example.flutter_tencent_upload.videoupload.TXUGCPublishTypeDef;

import java.util.HashMap;
import java.util.Map;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.EventChannel;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;

/**
 * FlutterTencentUploadPlugin
 */
public class FlutterTencentUploadPlugin implements FlutterPlugin, MethodCallHandler, EventChannel.StreamHandler {

    /// The MethodChannel that will the communication between Flutter and native Android
    ///
    /// This local reference serves to register the plugin with the Flutter Engine and unregister it
    /// when the Flutter Engine is detached from the Activity
    private MethodChannel channel;
    private static Context mContext;
    private EventChannel eventChannel;
    public static EventChannel.EventSink mEventSink = null;

    @Override
    public void onAttachedToEngine(@NonNull FlutterPluginBinding flutterPluginBinding) {
        if (null == mContext) {
            mContext = flutterPluginBinding.getApplicationContext();
        }
        channel = new MethodChannel(flutterPluginBinding.getBinaryMessenger(), "flutter_tencent_upload");
        channel.setMethodCallHandler(this);
        /**
         * 回调监听通道
         */
        eventChannel = new EventChannel(flutterPluginBinding.getBinaryMessenger(), "flutter_tencent_upload_stream");
        eventChannel.setStreamHandler(this);
    }

    @Override
    public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {
        if (call.method.equals("uploadVideo")) {
            Map<String, String> videoInfo = (Map<String, String>) call.arguments;
            if (videoInfo != null) {
                if (mEventSink != null) {
                    mEventSink.success(0.0);
                }
                String token = videoInfo.get("token");
                String filePath = videoInfo.get("videoPath");
                String coverPath = videoInfo.get("coverPath");
                uploadVideo(mContext, token, filePath, coverPath, new OnLiteUploadListener() {
                    @Override
                    public void onSuc(String id, String url, String coverUrl) {
                        Map<String, Object> callback = new HashMap<>();
                        Map<String, String> data = new HashMap<>();
                        data.put("id", id);
                        data.put("url", url);
                        data.put("coverUrl", coverUrl);
                        callback.put("code", 1);
                        callback.put("msg", "上传成功");
                        callback.put("data", data);
                        result.success(callback);
                    }

                    @Override
                    public void onProgress(long uploadBytes, long totalBytes) {
                        double progress = 1.0 * uploadBytes / totalBytes;
                        mEventSink.success(progress);
                    }

                    @Override
                    public void onFail(int code, String msg) {
                        Map<String, Object> callback = new HashMap<>();
                        callback.put("code", code);
                        callback.put("msg", msg);
                        callback.put("data", null);
                        result.success(callback);
                    }
                });
            }
        } else if (call.method.equals("init")) {
            result.success(true);
        } else {
            result.notImplemented();
        }
    }

    private interface OnLiteUploadListener {
        void onSuc(String id, String url, String coverUrl);

        void onProgress(long uploadBytes, long totalBytes);

        void onFail(int code, String msg);
    }

    private void uploadVideo(Context context, String mCosSignature, String mVideoPath, String mCoverPath, final OnLiteUploadListener onLiteUploadListener) {
        TXUGCPublish mVideoPublish = new TXUGCPublish(context);
        // 文件发布默认是采用断点续传
        TXUGCPublishTypeDef.TXPublishParam param = new TXUGCPublishTypeDef.TXPublishParam();
        // 需要填写第四步中计算的上传签名
        param.signature = mCosSignature;
        // 录制生成的视频文件路径, ITXVideoRecordListener 的 onRecordComplete 回调中可以获取
        param.videoPath = mVideoPath;
        // 录制生成的视频首帧预览图，ITXVideoRecordListener 的 onRecordComplete 回调中可以获取
        param.coverPath = mCoverPath;
        param.enableHttps = true;
        mVideoPublish.publishVideo(param);
        mVideoPublish.setListener(new TXUGCPublishTypeDef.ITXVideoPublishListener() {
            @Override
            public void onPublishProgress(long uploadBytes, long totalBytes) {
                if (onLiteUploadListener != null) {
                    onLiteUploadListener.onProgress(uploadBytes, totalBytes);
                }
            }

            @Override
            public void onPublishComplete(TXUGCPublishTypeDef.TXPublishResult result) {
                if (onLiteUploadListener != null) {
                    if (result.retCode == 0) {
                        onLiteUploadListener.onSuc(result.videoId, result.videoURL, result.coverURL);
                    } else {
                        onLiteUploadListener.onFail(result.retCode, result.descMsg);
                    }
                }
            }
        });
    }


    @Override
    public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
        channel.setMethodCallHandler(null);
    }

    @Override
    public void onListen(Object arguments, EventChannel.EventSink events) {
        mEventSink = events;
    }

    @Override
    public void onCancel(Object arguments) {
        mEventSink.endOfStream();
    }
}
